### PR TITLE
add Joint Ammo Magazines A3

### LINKED ISSUE #108: Add Joint Ammo and Magazines (JAM)
GOAL: 
Provide common bullet and magazine classes for use by weapon mods in order to achieve better consistency and compatibility

TO DO:
- add common bullets (CfgAmmo classes)
- add magazines (CfgMagazines classes)
- add magazine groups
- provide macros as alternative to magazine groups so mod makers can use them
- configure vanilla weapons to use new magazines
- add support for AB, ACE3

TO NOT DO:
- create too many bullet and magazine variations

Source: https://github.com/robalo/mods/blob/master/asdg_jam/config.cpp
Relates to #43

### PR DESCRIPTION
**Pull request by commy2**
Add JAM.

With a PR open, even if WIP, it is more convenient to see the diffs and discuss the implementation. :~)

-close #108

### CODE REVIEW

**Comment by Drofseh** [comment]
Additionally, shouldn't everything inherit from a base magwell class so that mods can add on?
If I were to do this in my mod, it would overwrite the BI mags from CBA right?
```
    class CBA_762x51_M14 {
        My_Mod_Mags[] = {
            "5000Rnd_762x51_M14Mag"
        };
    };
```

Should be this:
CBA:
```
    class CBA_Magwell_Base;
    class CBA_762x51_M14 : CBA_Magwell_Base {
        BI_mags[] = {
            "20Rnd_762x51_Mag"
        };
    };
```
My_Cool_Mod:
```
    class CBA_Magwell_Base;
    class CBA_762x51_M14 : CBA_Magwell_Base {
        My_Cool_Mod_Mags[] = {
            "5000Rnd_762x51_M14Mag"
        };
    };
```
In Game:
```
    class CBA_Magwell_Base;
    class CBA_762x51_M14 : CBA_Magwell_Base {
        BI_mags[] = {
            "20Rnd_762x51_Mag"
        };
        My_Cool_Mod_Mags[] = {
            "5000Rnd_762x51_M14Mag"
        };
    };
```

**Comment by commy2** [comment]
The base class is pointless if empty.

**Comment by Drofseh** [comment]
As I understand it, if CBA does this:
```
    class CBA_762x51_M14 {
        BI_mags[] = {
            "20Rnd_762x51_Mag"
        };
    };
```

And then I do this in my own mod
```
    class CBA_762x51_M14 {
        My_Mod_Mags[] = {
            "5000Rnd_762x51_M14Mag"
        };
    };
```
It will be treating as if I am creating a new `CBA_762x51_M14` and completely overwrite the original CBA class. Only my `My_Mod_Mags` will appear in game, not `BI_mags`.

**Comment by commy2** [comment]
You're wrong about that. That is not what would happen. The class would contain both entries. Base classes only have to be referenced if they exist.

**Comment by robalo** [comment]
A RHS compatibility addon would look somewhat like this (not a complete working example):
```
class CfgMagazineWells {
    class CBA_556x45_STANAG {
        RHS_mags[] = {
            "rhs_mag_30Rnd_556x45_M855_Stanag",
            "rhs_mag_30Rnd_556x45_M855_Stanag_Tracer_Red",
            "rhs_mag_30Rnd_556x45_M855_Stanag_Tracer_Green",
            "rhs_mag_30Rnd_556x45_M855_Stanag_Tracer_Yellow",
            "rhs_mag_30Rnd_556x45_M855_Stanag_Tracer_Orange",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag_No_Tracer",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag_Tracer_Red",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag_Tracer_Green",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag_Tracer_Yellow",
            "rhs_mag_30Rnd_556x45_M855A1_Stanag_Tracer_Orange",
            "rhs_mag_30Rnd_556x45_Mk318_Stanag",
            "rhs_mag_30Rnd_556x45_Mk262_Stanag",
            "rhs_mag_30Rnd_556x45_M200_Stanag"
        };
    };
    class CBA_762x39_AK {
        RHS_mags[] = {
            "rhs_30Rnd_762x39mm",
            "rhs_30Rnd_762x39mm_tracer",
            "rhs_30Rnd_762x39mm_89",
            "rhs_30Rnd_762x39mm_U"
        };
    };
    class CBA_545x39_AK {
        RHS_mags[] = {
            "rhs_30Rnd_545x39_AK",
            "rhs_30Rnd_545x39_AK_no_tracers",
            "rhs_30Rnd_545x39_7N6_AK",
            "rhs_30Rnd_545x39_7N10_AK",
            "rhs_30Rnd_545x39_7N22_AK",
            "rhs_30Rnd_545x39_AK_green",
            "rhs_30Rnd_545x39_7U1_AK",
            "rhs_45Rnd_545x39_AK",
            "rhs_45Rnd_545x39_7N6_AK",
            "rhs_45Rnd_545x39_7N10_AK",
            "rhs_45Rnd_545x39_7N22_AK",
            "rhs_45Rnd_545x39_AK_green",
            "rhs_45Rnd_545x39_7U1_AK"
        };
    };
    class CBA_762x54R_SVD {
        RHS_mags[] = {
            "rhs_10Rnd_762x54mmR_7N1"
        };
    };
    class CBA_RPG7 {
        RHS_rockets[] = {
            "rhs_rpg7_PG7V_mag",
            "rhs_rpg7_PG7VL_mag",
            "rhs_rpg7_PG7VR_mag",
            "rhs_rpg7_OG7V_mag",
            "rhs_rpg7_TBG7V_mag",
            "rhs_rpg7_TYPE69_airburst_mag"
        };
    };
    class CBA_9x39_VSS {
        RHS_mags[] = {
            "rhs_20rnd_9x39mm_SP5",
            "rhs_20rnd_9x39mm_SP6",
            "rhs_10rnd_9x39mm_SP5",
            "rhs_10rnd_9x39mm_SP6"
        };
    };
    class CBA_556x45_MINIMI {
        RHS_boxes[] = {
            "rhs_200rnd_556x45_M_SAW",
            "rhs_200rnd_556x45_B_SAW",
            "rhs_200rnd_556x45_T_SAW"
        };
        RHS_pouches[] = {
            "rhsusf_100Rnd_556x45_M855_soft_pouch",
            "rhsusf_200Rnd_556x45_M855_soft_pouch",
            "rhsusf_100Rnd_556x45_soft_pouch",
            "rhsusf_200Rnd_556x45_soft_pouch",
            "rhsusf_100Rnd_556x45_M200_soft_pouch"
        };
    };
    class CBA_762x51_M14 {
        RHS_mags[] = {
            "rhsusf_20Rnd_762x51_m118_special_Mag",
            "rhsusf_20Rnd_762x51_m993_Mag",
            "rhsusf_20Rnd_762x51_m62_Mag"
        };
    };
    class CBA_40mm_M203 {
        RHS_grenades[] = {
            "rhs_mag_M441_HE",
            "rhs_mag_M433_HEDP",
            "rhs_mag_M781_Practice",
            "rhs_mag_M397_HET",
            "rhs_mag_M4009",
            "rhs_mag_m576",
            "rhs_mag_M585_white",
            "rhs_mag_M661_green",
            "rhs_mag_M662_red",
            "rhs_mag_M713_red",
            "rhs_mag_M714_white",
            "rhs_mag_M715_green",
            "rhs_mag_M716_yellow"
        };
    };
    class CBA_40mm_GP {
        RHS_grenades[] = {
            "rhs_VOG25",
            "rhs_VOG25p",
            "rhs_vg40tb",
            "rhs_vg40sz",
            "rhs_vg40op_white",
            "rhs_vg40op_green",
            "rhs_vg40op_red",
            "rhs_GRD40_white",
            "rhs_GRD40_green",
            "rhs_GRD40_red",
            "rhs_VG40MD_White",
            "rhs_VG40MD_Green",
            "rhs_VG40MD_Red",
            "rhs_GDM40"
        };
    };
};

class cfgWeapons {
    class rhs_weap_m4_Base : arifle_MX_Base_F {
        magazineWell[] = {"CBA_556x45_STANAG"};
    };
    class rhs_weap_ak74m_Base_F : Rifle_Base_F {
        magazineWell[] = {"CBA_545x39_AK"};
    };
    class rhs_weap_svd : rhs_weap_ak74m {
        magazineWell[] = {"CBA_762x54R_SVD"};
    };
    class rhs_weap_akm : rhs_weap_ak74m {
        magazineWell[] = {"CBA_762x39_AK"};
    };
    class rhs_weap_rpg7 : Launcher_Base_F {
        magazineWell[] = {"CBA_RPG7"};
    };
    class rhs_weap_asval : rhs_weap_ak74m {
        magazineWell[] = {"CBA_9x39_VSS"};
    };
    class rhs_weap_saw_base : Rifle_Base_F {
        magazineWell[] = {"CBA_556x45_MINIMI", "CBA_556x45_STANAG"};
    };
    class rhs_weap_m240_base : rhs_weap_M249_base {
        magazineWell[] = {};
    };
    class rhs_weap_sr25 : rhs_weap_m14ebrri {
        magazineWell[] = {};
    };
    class rhs_weap_M320_Base_F : Pistol_Base_F {
        magazineWell[] = {"CBA_40mm_M203"};
    };
    class GP25_Base : UGL_F {
        magazineWell[] = {"CBA_40mm_GP"};
    };
    class rhs_weap_m70ab2 : rhs_weap_m70_base {
        magazineWell[] = {"CBA_762x39_AK"};
    };
    class rhs_weap_m92 : rhs_weap_m70_base {
        magazineWell[] = {"CBA_762x39_AK"};
    };
};
```

File: addons/jam/CfgMagazineWells.hpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    class CBA_545x39_AK {
+        BI_mags[] = {
+            "30Rnd_545x39_Mag_F",
+            "30Rnd_545x39_Mag_Green_F",
+            "30Rnd_545x39_Mag_Tracer_F",
+            "30Rnd_545x39_Mag_Tracer_Green_F"
+        };
+    };
+    class CBA_556x45_MINIMI {
+        BI_boxes[] = {
```
Comment (PabstMirror) [inline addons/jam/CfgMagazineWells.hpp]: Add the 30-round STANAG mags as well?

**Comment by jonpas** [comment]
What about mag types? It's something @dedmen brought up. Some mags housing same caliber but are of different design and/or capacity might not fit.

**Comment by Drofseh** [comment]
> Some mags housing same caliber but are of different design and/or capacity might not fit.

How deep down the rabbit hole do you want go? Even weapons that are nominally STANAG compliant can have problems with some magazines.
For example, PMAGs don't work properly in the HK-416/M27, or the F2000.

**Comment by jonpas** [comment]
> How deep down the rabbit hole do you want go?

I don't know and don't really care, just putting it out there because it was a concern on Slack the other day.

**Comment by robalo** [comment]
My first thought was splitting the "STANAG"s might be too much. But I wouldn't mind going 100% realistic about it provided some solid info.

**Comment by dedmen** [comment]
In my implementation I've split stanag small/big. Because sniper rifles only take 5/10rnd mags and not all weapons support 100/150rnd mags. Then left it to the weapon class to include both magwells.
https://gist.github.com/dedmen/dc980cf8d6b4b719484e2dd8158a7de1

**Comment by robalo** [comment]
See, that's the thing, info on that might be hard to find or debatable. For example, why didn't you allow drum mags for HK33 in your implementation when things like this exist ? https://www.xproducts.com/product/x33-hk53-drum-magazine-for-hk33-hk53/
It might be best to avoid these issues perhaps.

**Comment by Drofseh** [comment]
My vote would be to keep it reasonably simple. If the weapon and the magazine are at least generally STANAG compliant then have them all be interchangeable. Even for things like the PMAG and the M27.

**Comment by robalo** [comment]
* agree
* SG556 uses M16 mags; why ?
* just for readability
* right, doesn't hurt anything tho

**Comment by Blutze** [comment]
Meant the 7.62x51 variant from the Marksman DLC. Kinda like that little thing, so I just wanted to know which mags it might end up getting. With RHS M993 for instance, its <0.6 MoA.

**Comment by commy2** [comment]
WIP as long as BI doesn't fix the internals of this.

**Comment by dedmen** [comment]
Maybe with Encore.. Atleast RHS said they are pushing their update after Encore. Maybe they are waiting for this.

**Comment by Dahlgren** [comment]
12.7mm (.50 cal) magazines for snipers would be nice, such as the GM6 Lynx in Arma 3 or Barret rifles in many mods

**Comment by robalo** [comment]
Weird, BIKI says mag wells are available since 1.80: https://community.bistudio.com/wiki/Arma_3_Weapon_Config_Guidelines

**Comment by dedmen** [comment]
@robalo confirmed. But we are mostly waiting for BI's implementation of magazine wells to see what we actually have to implement and what's already there. We should use the vanilla magazine wells if they already exist instead of creating new ones.
Should be there now with Encore update though.

**Comment by robalo** [comment]
I doubt BI will ever add any magwells config data because they have no reason to do it.
I'll unpack and check 1.84 tonight.

**Comment by robalo** [comment]
Sounds good. Nothing configured in 1.84. I guess they'd be doing it for 3rd party DLC.

**Comment by dedmen** [comment]
RHS will atleast implement them. Not sure if it would be worth to wait and see how they did it.

**Comment by dedmen** [comment]
`But if we don't push this sooner they have nothing to consider.` Their implementations are already done. They have nothing to consider no matter when we push this. But if we push this without keeping the vanilla implementation in mind we'll have 2 competing standards out of which one is in vanilla Arma.
We can add on what vanilla is missing. But we shouldn't push duplicates.

**Comment by robalo** [comment]
RHS never cared much for vanilla compatibility, their mags didn't work in vanilla weapons nor viceversa.
Regardless of what they and BIS do, we have a broader scope to cover. Are you suggesting we build upon whatever RHS "implementation" is and then keep up with it ? Also vanilla "implementation" is just in the engine, plus VA support, they have no reason to go beyond that. If I waited for them to implement compatibleitems classes in the past, JR would've never happened,

**Comment by dedmen** [comment]
>RHS never cared much for vanilla compatibility

that was before RHS had a BI developer in their team.

>Are you suggesting we build upon whatever RHS "implementation" is and then keep up with it ?

That "whatever" is the vanilla implementation that we will get in dev-branch soon:tm: and in the next DLC.

>Also vanilla "implementation" is just in the engine

No. The Vanilla CfgMagazineWells config is what I mean by implementation.

>they have no reason to go beyond that

But they are doing it already.

**Comment by robalo** [comment]
Some good 2 know infos. Guess we'll wait and see. Thanks !

**Comment by Blutze** [comment]
Something didn't add up about dedmen's argumentation, and it just hit me (good ol' showerthoughts): A significant portion of the RHS weapons have no equivalent in vanilla. They are gonna have to introduce class names for Remington 700s (two different ones even), PKs, all the 9x39mm stuff, and a variety of others. For those, other mod makers will once again have to create compatibility code between RHS and the rest of the world.

Is it really so hard for them to contribute to this PR?

**Comment by dedmen** [comment]
>A significant portion of the RHS weapons have no equivalent in vanilla. They are gonna have to introduce class names for Remington 700s (two different ones even), PKs, all the 9x39mm stuff, and a variety of others. For those, other mod makers will once again have to create compatibility code between RHS and the rest of the world.

Correct. These are the things we will ALSO implement then. Sadly. RHS doesn't require CBA. But I guess they could also add their magazines to our Well's.

Waiting for answer from reyhard.

**Comment by Blutze** [comment]
Silent native compatibility would make it easier on the end user, not having to download a bunch of extra mods. It doesn't however help at all with the development cycle issues behind compats: Every time RHS changes something, like adding a new mag or depreciating an old one, the mod offering compatibility is out of date for a while. Sure, CBA is a bit more organized and a bit more functionality/coding focussed than NIArms, but those compat mods have been missing the 0.4.3 mags for almost a year now.

In the end, there is only one acceptable way of tackling this. CBA extends the vanilla 2035 magwells into a 2018 JAM standard, everybody else implements that standard. "Everybody" including CUP, SMA, NIArms, RHS and others _as peers_.

**Comment by Dahlgren** [comment]
@dedmen sadly the game throws errors/warnings for magazine declarations inside CfgMagazineWells which are not available

**Comment by dedmen** [comment]
Here are the RHS/Vanilla magazine wells.
https://gist.github.com/dedmen/15e96189bd4c8edc148b136686275865
The naming for them looks all the same for vanilla/rhs. So I'd say we can just assume that naming scheme is vanilla.

**Comment by dedmen** [comment]
Copy-paste of what I wrote to fuxna

>One big problem that we were debating about is magazine size. For example you have a Stanag_556x45 magazineWell.
But there are everywhere from 10 to 150 round magazines. And if you only have one big Stanag group one can't take care of weapons that only take small magazines and can't incorporate the big ones.
People would have to make their own weapon specific magazineWells which then won't be compatible with other mods.
I see that problem is also not fixed in the CBA magazineWells yet. I'll get some more attention onto that

As example of that just look at https://github.com/CBATeam/CBA_A3/pull/928/files#diff-ddb5c69e174931ad1c5d4a5d93f8a28bR19
This magazineWell is basically useless for every weapon that cannot fit 150rnd drum mags for whatever reason.
Maybe we should split them like 5/10/20/30/50/100/150 round magazines?
Or maybe once small/20/30/big? If you can fit a 100rnd you can probably also fit a 150rnd magazine. So we can just combine them as `big`
5/10 is a rather big difference though. So maybe 5/10/20/30/big ?
Ofcause we end up with a lot of mag groups. But modularity doesn't hurt us here. Modularity should be high up on the priority list.

**Comment by dedmen** [comment]
What is a "Coffin" magazine?
I agree that's a better solution. Is there any reason why a gun would be magazine length limited too?
Like I could see some Sniper rifles that take a Stanag 7,62 magazine but obviously don't want it to take 30 round magazines.
Would we want to again split it apart and for at least the relevant calibers (7,62mm/.300/.308/.338/.50)?
Split the "Stick" ones up between 5/10/20/bigger?
Splitting up Drum/Coffin(whatever that is :D) obviously wouldn't make sense as weapons that are limited by magazine "length" can't fit these anyway.

**Comment by Blutze** [comment]
http://www.imfdb.org/images/7/7d/Avengers3_241.jpg This would be a 100 round coffin. Also known as quadstack. Difference between these and drums is width - they only clip with the weapon model, not the user.

Length is a consideration when prone on the ground and because varying follower spring pressure over the duration of a mag could cause differences in how the rounds end up in the chamber, shifting your point of impact. 20 rounds for marksman rifles and 10 (or less) rounds for sniper rifles should be small enough for the latter not not be a problem, and the former is solved by drums being relatively short (often shorter than the standard size stick) plus not really modelled in the game anyway. The only actually longer 7.62x51 mag I am aware of is the OSW FAL 30 rounder - everything else is either 20 or a drum.

**Comment by commy2** [comment]
Just make something. People will complain, and then you can adjust. And now you can even blame them back for not participating when they had a chance to make the initial decissions.

**Comment by bux** [comment]
> Just make something.

👍 

> Shipping beats Perfection

**Comment by dedmen** [comment]
I agree. This PR right now has all combined magazine sizes. We could already merge this and then just add size-splits later.

**Comment by robalo** [comment]
It's good to go as far as I'm concerned, I'm no big fan of making it ultra realistic but don't mind going that way too if there's high demand and good info sources.

**Comment by Drofseh** [comment]
My vote is to merge now.

In future, the most I would want to see a magwell split into sub groups would be
`MyMagWell_Standard` - 'normal' capacity or anything physically smaller (eg, for STANAG 30 round magazines or smaller.
`MyMagWell_Large_Stick` - any stick, coffin, or other magazine that is physically larger than `Standard` (eg, for AKs the 45 round RPK magaines, for STANAG the Surefire MAG5-60.
`MyMagWell_Drum` - all drum magazines

**Pull request merged**